### ISSUE TITLE
API for dynamic text

### ISSUE THREAD

**Issue by maximilianriemensberger** [alengwenus/pypck#43]
The API for dynamic text suffers from str and byte string encoding issues.  Right now module.py `dyn_text` takes a string encode it as bytes (UTF-8 characters to bytes conversion), splits it up into up to 5 parts each at most 12 bytes. Those parts are handed to PckGenerator `dyn_text_part` to generate the command. The command is returned as `str` (not byte str). connection.py `send_command` encodes the `str` into bytes again.

For the whole thing to work, we would actually need the generator to return bytes (encoding is already done) or the dyntext parts need to be added later in the send_process after the conversion tool place.  Otherwise we would need to handle partial UTF-8 bytes in python `str`, which is impossible.

**Comment by alengwenus** [comment]
Since for Python3 all strings are encoded in unicode, we should be able to use strings in the whole process up to  `PchkConnectionManager.send_command` which then has to do the encoding to bytes for writing to the socket.
But we have to assure that the string parts which are passed to `PckGenerator.dyn_text_part` don't have a byte size >12.

I guess the split-logic in `ModuleConnection.dyn_text` has to be rewritten in a way that the corresponding bytes of the unicode characters are not separated.

**Comment by maximilianriemensberger** [comment]
Your proposal assumes that LCN is ok with sending for example 11 bytes and 7 bytes and it results in the same displayed text as if sending 12 and 6 bytes. If this is the case, then I agree it should work. It has only one downside, that you might waste a few bytes as you cannot fully exploit the 12 bytes in each part. But I don't think this is major issue since the number of wasted bytes will most of the time not exceed 6 (3 unlucky splits at three byte UTF-8 chars with 2 bytes lost each).

The parts breaking logic will be a bit ugly but that's no big thing either :smirk:.

**Comment by alengwenus** [comment]
Unfortunately I am not able to test it because I don't have a text display... Just guessing...

I am thinking about something like this:
```python3
def split_text(text):
    size = 0
    parts = [""] * 5
    for char in text:
        size += len(char.encode())
        part_id = (size - 1) // 12
        if part_id > 5:
            raise ValueError("Text to long (max. 60 bytes).")
        parts[part_id] += char

    return parts
```

**Comment by alengwenus** [comment]
Maybe we can use LCN-Pro's bus monitor to test...

**Comment by maximilianriemensberger** [comment]
Sure. I can test it  with a display this weekend.

**Comment by alengwenus** [comment]
FYI: PCHK responds the command `GTDT11AAAAAAAAAAAé` (12 chars, but 13 bytes) with `(GTDT textlen=13 valid: 0-12 chars)`.
So I guess we are right...  Unforntunately LCN-PRO does not show the messages (only row and part information).

**Comment by maximilianriemensberger** [comment]
Also I think the split_text needs to be a bit more refined:

```
In [6]: split_text('aüüüüüüüüüüüaüüüüüüüüüüüü')
Out[6]: ['aüüüüü', 'üüüüüüa', 'üüüüüü', 'üüüüüü', '']
```

**Comment by alengwenus** [comment]
May be [this](https://stackoverflow.com/questions/6043463/split-unicode-string-into-300-byte-chunks-without-destroying-characters) helps...

**Comment by maximilianriemensberger** [comment]
Slight variation works:
```py
def split_utf8(s, n):
    """Split UTF-8 s into chunks of maximum length n."""
    b = s.encode('utf-8')
    while len(b) > n:
        k = n
        while (b[k] & 0xc0) == 0x80:
           k -= 1
        yield b[:k].decode('utf-8')
        b = b[k:]
    yield b.decode('utf-8')
```

**Comment by maximilianriemensberger** [comment]
@alengwenus I quickly play around with dynamic text on a GT4D.  I looks not as straight forward as we thought. For example sending `GTDT31asdfasdfasd` indeed prints `asdfasdfasd`. Sending just `GTDT31asdfasdfasdf` prints nothing but the module sends a pos ack. Sending `GTDT31asdfasdfasd` and later `GTDT32asdf` only prints the former (both are pos acked). I think I didn't try `GTDT31asdfasdfasdf` and `GTDT32asdf` before I had to leave. I retry later to figure out the exact semantics and report back.

**Comment by maximilianriemensberger** [comment]
I continued with the experiments (not yet unicode, because it's somewhat difficult to type partial utf-8 chars into telnet).  But with pure ascii I found the following behavior:

* `GTDT31asdfasdfasd` prints `asdfasdfasd`. This is the case regardless of anything I send in `GTDT32...`, `GTDT33...`, etc.
* `GTDT31asdfasdfasdf` prints nothing so far, when later issuing `GTDT32` (empty string) it prints `asdfasdfasdf`.
* `GTDT31asdfasdfasdf` plus `GTDT32asdf` prints `asdfasdfasdfasdf`
* `GTDT310123456789ab` plus `GTDT32cdef01234567` prints nothing so far and followed by `GTDT33` (empty string prints) `0123456789abcdef01234567`. If followed by `GTDT338` instead it prints `0123456789abcdef012345678` and so on.
* `GTDT31` (empty string) clears the display.

So I believe that every parts need to be filled up with 12 bytes before the next part is started. Also I think all parts should probably be sent every time the dyn text is set. 

Another observation is that parts can be individually updated and are shown if no earlier part terminates the string by being shorter than 12 bytes.

Also I think that we have to do a strict 12 bytes split also for non-ascii utf-8. Otherwise, the module might detect a premature end of the dyn text like in the ascii examples. This has a few implications:
* The parts cannot be represented by python `str`. I would recommend to change the `PckGenerator.dyn_text` to return either `Tuple[str, bytes]` or just `bytes`. In the former case the header would continue to be a string, in the latter case `PckGenerator.dyn_text` fully encodes the command to bytes. In any case, everybody need to special case dyn_text. The alternative is to switch completely to bytes for `PckGenerator`. Also not really pretty.
* When you have a dyn_text currently show that has partial utf-8 chars at boundaries, switching the content risks that for a brief period of time the module might have a garbage utf-8 char at such a boundary. Since I have read quite a few threads about bricked (at least temporarily) displays, I'd prefer to have the safety option go through an intermediate step, where the previous text is cleared before the new text is shown. This could either mean sending 5 empty parts before the new text is sent or sending just an empty first part  and then the other parts backwards such that the text is only show once all parts are here. But I have no idea if this actually works.

Long story short: I think I'll prepare a branch and a set of test strings to test some of these things in particular with unicode.

**Comment by alengwenus** [comment]
Thanks a lot for testing!
The behavior makes sense and could be explained as follows: When analyzing the plain bus-telegrams, all bytes of the text parts are preallocated with 0x00 (C-string termination character). So I guess the display puts together all 5x12byte code blocks and interprets the whole string until a 0x00 char is reached, no matter if it is followed by any other character from previous messages. Sent blocks will just be replaced in the displays internal buffer.
Now, it would be interesting what happens at the borders with two-byte unicode chars... I guess you can break them apart in the PCK messages (which might look quite strange for the PCK commands).
If that's the case, I would suggest to put all the logic into the `AbstractConnection`'s convenience method and break the text into 12 byte parts (->string length possibly < 12). This might cause unicode characters to be broken, but so what? Then we send the 12 byte strings to the `PckGenerator.dyn_text`, possibly followed by an empty text block if the byte length is a multiple of 12.

**Comment by maximilianriemensberger** [comment]
Your analysis very plausibly explains the behavior I observed. I think the current `dyn_text` code in `AbstractConnection` is quite close to what it needs to do. It already breaks exactly at 12 bytes. I just need to carry the byteness of it all the way through to the final `write()` call.

I've a branch that drafts how this might work. Still working on the tests.
